### ISSUE TITLE
Easy way to get HTML into an event title

### ISSUE THREAD

**Issue by blag001** [fullcalendar/fullcalendar#2919]
Hi there,
I'm using fullcalendar (nice work~) and find what it seem to be a missing function.
When you put html in the event title, it'll always be html-escaped, I understand it's the best default way, as some title may contain some html char and user input that should be bulletproof .

But why you don't have put an `(bool) textEscape {default : TRUE}` option in [Event Source Object](http://fullcalendar.io/docs/event_data/Event_Source_Object/#options) ?
(and in [Event Object](http://fullcalendar.io/docs/event_data/Event_Object/) to as it's the same)

I'm a big boy, I'm able to deal with htmlEscape by myself and I think I'm not the only one here... (i've found many issue about html in title on google)
But having to use an ugly hook with eventRender just to put a span for a glyphicons seem to be a bit overkill for me...

Tanks,
Blag

TL;DR : 
- adding option `(bool) textEscape {default : TRUE}` on `Event Source Object` and `Event Object`
- dealing with this kind of data : 

``` json
{
    "events": [
        {
            "title": "<b>Event in bold</b> 1",
            "start": "2011-04-04"
        },
        {
            "title": "<span class=""glyphicons""></span> Event with glyphicons",
            "start": "2011-05-05"
        }
        // etc...
    ],
    "textEscape": false,   // new option to disable the auto escape
    "color": "yellow",   // an option!
    "textColor": "black" // an option!
}
```

or 

``` json
{
    "events": [
        {
        "title": ">> Event  escaped",
        "start": "2011-04-04"
        },
        {
            "title": "<span class=""glyphicons""></span> Event with HTML !",
            "start": "2011-05-05",
            "textEscape": false   // new option to disable the auto escape
        }
        // etc...
        ],
    "color": "yellow",   // an option!
    "textColor": "black" // an option!
}
```

**Comment by rePhat** [comment]
You can actually use http://fullcalendar.io/docs/event_rendering/eventRender/ to render your HTML title. Thats the way I'm doing it (also for adding icons). Hope this helps :)

**Comment by blag001** [comment]
Yes, I use it to, like I said : `But having to use an ugly hook with eventRender just to put a span for a glyphicons seem to be a bit overkill for me...`

I just ask to make an official way to put html, without triggering a hook on every event just to put a span on 3 of them...

Forcing escape is like forcing the php magic quote for me, useful for beginner [I'll not say noob but nearly] but such a pain when your are a bit more skilled and able to deal with it by yourself. 
And if we take a look, even PHP now disable default magic-quote.

**Comment by arshaw** [comment]
the issue i have with this is more of a philosophical one. if you include HTML in your event data, you are all of a sudden mixing your data layer with your presentation layer.

I'd rather tackle #1009 to help you achieve what you want, specifically for icons. more semantic

**Comment by blag001** [comment]
HTML is structural, not a presentation layer at all (this is CSS), it's data on data, it's like saying meta-data are not data...

So yes, it's **should** be allowed for user who want to use it in title (and the use of the option like I suggest, by default, won't change anything).

I work with [CakePhp framework](https://en.wikipedia.org/wiki/CakePHP), and every helper that auto-escape content allow you to disable-it. If you need to tell navigator that a part of the link is more/less important than the other, you'll need to put html in it and hopefully you'll not have to make a JS hook to change it on the fly (which is against every good SEO).
I think it's the same in JS, you shouldn’t need to make a second change on the fly to get your final data (and I *really* don't like the performance cost of this second fly to be honest).

BTW, if we speak about mixing data layer with presentation layer on **Event Object** : `textColor`, `borderColor`, `backgroundColor` and `color` are way more against it that a bit of structural html...

Regards,
blag

**Comment by arshaw** [comment]
You bring up a good point. The thing that weirds me out is the meta-data aspect. The event data containing another type of data that is specific to one particular type of rendering engine (a web browser). Maybe I'm over-abstracting, but it'd be cool if the FullCalendar event data format was something beyond what a JavaScript plugin needs. Similar to what iCalendar aims to be. It could be accepted by an iOS implementation of FullCalendar for example (a different type of rendering engine). In that case this alternative client could make use of presentational properties, like `textColor` but interpretation is up to the client. It would be awkward to try to render HTML in a non-browser setting. Maybe iOS has these capabilities, but beside the point.

I was curious what the iCalendar spec does. They do not offer a standard way to specify HTML within an event title or description. However, a "calendar extension" seems to have evolved to accommodate this. See `X-ALT-DESC`:
https://en.wikipedia.org/wiki/ICalendar

I'm all about being pragmatic however... if people are struggling to get around this limitation by introducing some other type of non-standard property with HTML in it, that defeats the point. And implementing this is pretty easy. I'll get this in an upcoming release.

**Comment by arshaw** [comment]
this will likely be another property name altogether, rather than a flag. will probably be `titleHtml`

**Comment by blag001** [comment]
Ok, I now understand better your point (agnostic data).
But yes, as you are working on a JS implementation, I think it's not a good idea to avoid an html support.
`titleHtml` seem nice to me, it's like the way you deal with email (one content in html and another in plain/txt)
Thanks again,
Blag

**Comment by Narzerus** [comment]
Any updates on this?

**Comment by perkins2099** [comment]
Also curious about this one. :)

**Comment by gusreyes01** [comment]
This function would be useful to me too.

**Comment by arshaw** [comment]
no update, but if you need a workaround, `eventRender` is there for you

**Comment by trangchuvn** [comment]
eventRender: function( event, element, view ) {
       element.find('.fc-title').prepend('<span class="glyphicon"></span> '); 
}

**Comment by tyrann0us** [comment]
If you already got HTML in the title you could do something like this:

```
eventRender: function( event, element, view ) {
 	var $title = element.find( '.fc-title' );
	$title.html( $title.text() );
}
```

**Comment by githubjeka** [comment]
`eventRender` doesn't work for list views

**Comment by githubjeka** [comment]
```js
eventRender: function( event, element, view ) {
        var title = element.find('.fc-title'); 
        var titleList = element.find( '.fc-list-item-title');            
        title.html(title.text());
        titleList.html(titleList.text());
},
```

**Comment by MartijnWelker** [comment]
In my opinion the nicest way to do this is to leverage the custom views. This is my code I use to do this in the scheduler plugin.

```
		var viewClass = FC.views.timeline.resourceClass;
		
		FC.views.customTimeline = viewClass.extend({
			initialize: function () {
				viewClass.prototype.initialize.apply(this, arguments);
				
				this.timeGrid.fgSegHtml = fgSegHtml;
			}
		});
		
		function fgSegHtml (seg, disableResizing) {
			var event = seg.event,
				isDraggable = this.view.isEventDraggable(event),
				isResizableFromStart = seg.isStart && this.view.isEventResizableFromStart(event),
				isResizableFromEnd = seg.isEnd && this.view.isEventResizableFromEnd(event),
				classes = this.getSegClasses(seg, isDraggable, isResizableFromStart || isResizableFromEnd),
				timeText = this.getEventTimeText(event);
			
			classes.unshift('fc-timeline-event', 'fc-h-event');
			
			return '<a class="' + classes.join(' ') + '" style="' + FC.cssToStr(this.getSegSkinCss(seg)) + '"' + (event.url ? ' href="' + FC.htmlEscape(event.url) + '"' : '') + '>' +
				'<div class="fc-content">' +
				(timeText ? '<span class="fc-time">' + FC.htmlEscape(timeText) + '</span>' : '') +
				'<span class="fc-title">' + (event.title ? event.title : '&nbsp;') + '</span>' +
				'</div>' +
				'<div class="fc-bg" />' +
				(isResizableFromStart ? '<div class="fc-resizer fc-start-resizer"></div>' : '') +
				(isResizableFromEnd ? '<div class="fc-resizer fc-end-resizer"></div>' : '') +
				'</a>';
		}
```

**Comment by nicolas15000** [comment]
Hello, this is my rendering function 👍 

`function getAllRendezVous(){
		
		$scope.eventSources.splice(0, $scope.eventSources.length);// REINIT TOTAL DES RENDEZ VOUS
		
		rdvFactory.get_rdvs().then(function(data){  // On récupère en Ajax tous les rendez-vous.
			
			var temp = data.data;
			
			angular.forEach(temp,function(value,index){
				   temp[index]["title"] = temp[index]["title"]+'<span> <i class="fa fa-users" aria-hidden="true"></i> -- Juristes:'+temp[index]["nbjuristes"];
				  temp[index]["className"] = "evenement2";
				  /* GESTION DE LA COULEUR DES RENDEZ VOUS EN FONCTION DE LA CRITICITE */
				  if(temp[index].criticite==5){
					  temp[index]["color"] = "red";
					 /*  temp[index]["className"]= 'fa fa-shopping-basket'; */
				  }
				  if(temp[index].criticite==4){
					  temp[index]["color"] = "orange";
				  }
				  if(temp[index].criticite<4){
					  temp[index]["color"] = "#3987ad";
					 
				  }
				  
			}); 
		
		$scope.eventSources = angular.copy([data.data]); 	
		
		})
	}`

**Comment by billfranklin** [comment]
status of this?

**Comment by fra-gadio** [comment]
@billfranklin  'trangchuvn' and 'tyrann0us' solutions are working for me.

**Comment by maieul** [comment]
It seems me that it should be a native feature, as @blag001 explained.

**Comment by matthewford** [comment]
I needed to do this in eventAfterRender

**Comment by tcurtin** [comment]
As a comment - the suggested `element.find('.fc-title')` does **not** work for list views, which have class `fc-list-item-title ui-widget-content` instead of `fc-title ui-widget-content`.  The following will find those as well:

`element.find('[class*=-title]')`

(the *= is important because the class selectors don't work on individual classes as you might expect, but the full list of classes as a string, so $= will fail.)

**Comment by Ahmad2323** [comment]
HI everyone, i am facing issue in fullcalendar. when i fetch the data from db , events.push event assighning undefined value to the fields. while json data come from db in correct format. Please help me to sort out this issue. Thanks

var events = [];
$.ajax({
url: "schedule.aspx/GetDataCustomer",
type: "POST",
dataType: "json",
cache: false,
contentType: "application/json; charset=utf-8",
success: function (data) {
$.each(data, function (i, v) {
events.push({
id: v.sch_id,
title: v.Subject,
description: v.Description,
start: moment(v.Start).format('YYYY-MM-DD'),
end: moment(v.End).format('YYYY-MM-DD'),
Lunch_S: v.Lunch_Start,
Lunch_E: v.Lunch_End,
color: v.ThemeColor,
allDay: v.IsFullDay
});
})

**Comment by amitshahc** [comment]
using `html: true` solved my problem.
in boostrap popover attribute use `data-html="true"`
```
eventRender: function (eventObj, $el) {
                    $el.popover({
                        title: eventObj.title,
                        content: eventObj.description,
                        trigger: 'hover',
                        placement: 'top',
                        container: 'body',
                        html: true
                    });
                }
```

**Comment by paulpgm** [comment]
I need HTML in the title to support multiple links within it. It would me to have the titleHTML feature implemented.
I implemented this in v4 by adding htmlTitle to the events within my json feed. This creates an extended property which I then place in the element while rendering.

```
eventRender: function(info) {
	if ("htmlTitle" in info.event.extendedProps)
		info.el.firstChild.innerHTML = info.event.extendedProps.htmlTitle;
}
```

**Comment by BernhardK91** [comment]
> ```
> eventRender: function(info) {
> 	if ("htmlTitle" in info.event.extendedProps)
> 		info.el.firstChild.innerHTML = info.event.extendedProps.htmlTitle;
> }
> ```

I have used the Idea from @paulpgm but there was a problem, wenn opening the list-view (wrong column got the title). So I used @paulpgm 's idea can extended it a little bit:

```
if ("htmlTitle" in info.event.extendedProps) {
    if(info.el.getElementsByClassName('fc-title').length > 0) {
        info.el.getElementsByClassName('fc-title')[0].innerHTML = info.event.extendedProps.htmlTitle;
    }
    if(info.el.getElementsByClassName('fc-list-item-title').length > 0) {
        info.el.getElementsByClassName('fc-list-item-title')[0].innerHTML = info.event.extendedProps.htmlTitle;
    }
}
```

Perhaps it helps anybody. :-)

**Comment by pepijndik** [comment]
@BernhardK91 
I am trying to add Text to my Title/ Add some things 
![20190903_121126](https://user-images.githubusercontent.com/23504693/64183262-eb88a680-ce69-11e9-87a3-aefa3a9a77dd.jpg)

Tried this:
```javascript
info.el.getElementsByClassName('.fc-title').append("<br/>" + info.event.extendedProps.klantnaam);
```

and Tried your example
```javascript
 if (info.view.type === "resourceTimeGridDay") {

            if ("klantnaam" in info.event.extendedProps) {
              if (info.el.getElementsByClassName('fc-title').length > 0) {
                info.el.getElementsByClassName('fc-title')[0].append = info.event.extendedProps.klantnaam;
              }
              if (info.el.getElementsByClassName('fc-list-item-title').length > 0) {
                info.el.getElementsByClassName('fc-list-item-title')[0].append = info.event.extendedProps.klantnaam;
              }
            }
          }
```
But still I don't now how I can do it the way I wanted.
hope you can explain me.

**Comment by HarveyEV** [comment]
Yeah so I had the same issue inserting icons in the event title. I just passed a class to the parent fc-event and created a pseudo-element for the class. Worked like a charm after screwing with the layout. No event render required.

**Comment by NahuBocco** [comment]
Hi! if you are using Angular, try this:

`eventRender(e){
    e.el.querySelectorAll('.fc-title')[0].innerHTML = e.el.querySelectorAll('.fc-title')[0].innerText;
  }`

and in your HTML

`<full-calendar (eventRender)="eventRender($event)"></full-calendar>`

**Comment by deepkesh26** [comment]
> I need HTML in the title to support multiple links within it. It would me to have the titleHTML feature implemented.
> I implemented this in v4 by adding htmlTitle to the events within my json feed. This creates an extended property which I then place in the element while rendering.
> 
> ```
> eventRender: function(info) {
> 	if ("htmlTitle" in info.event.extendedProps)
> 		info.el.firstChild.innerHTML = info.event.extendedProps.htmlTitle;
> }
> ```

It works for me I have modified a little bit for my usage.

 eventRender: function (info) {
                    if ("htmlTitle" in info.event.extendedProps) {
                        info.el.setAttribute("title",info.event.extendedProps.htmlTitle);
                    }
                },

**Comment by jayakrishnan-m** [comment]
Try eve.el.innerHTML = eve.el.innerText; .
Works on angular

**Comment by arshaw** [comment]
as of v5 ([which has just been released](https://fullcalendar.io/blog/2020/06/v5-official)), there's a much easier way to do this now.

demo: https://codepen.io/arshaw/pen/ExPXjpY?editable=true&editors=001

useful links:
https://fullcalendar.io/docs/event-render-hooks
https://fullcalendar.io/docs/content-injection

**Comment by pepijndik** [comment]
I saw it super easy.Really nice that you have implemented this feature!!I love it!!Op 24 jun. 2020 16:39 schreef Adam Shaw <notifications@github.com>:
as of v5 (which has just been released), there's a much easier way to do this now.
demo: https://codepen.io/arshaw/pen/ExPXjpY?editable=true&editors=001
useful links:
https://fullcalendar.io/docs/event-render-hooks
https://fullcalendar.io/docs/content-injection

**Comment by Cris200992** [comment]
But now on v5 if you change with eventcontent and you have a lot of resources the text will be out of event. could we change it like with event render as other versions?
![image](https://user-images.githubusercontent.com/35325847/88764603-8589c980-d175-11ea-8a4e-c869a5d82e1a.png)

**Comment by pepijndik** [comment]
@Cris200992  
I have got the same issue  
Used the Vertical Resource view on a large screen it looks ok. but when the screen is smaller Characters are getting out of the events. but when i used the old way of settings text into a event the event prop gets resized.

In the pictures i have blacked out user data like names
Smaller screeen pic: 

![Planning_example](https://user-images.githubusercontent.com/23504693/88766071-b408a400-d177-11ea-9c5d-e5ca7a43e83a.PNG)

Larger screen:
![Planning_example_@](https://user-images.githubusercontent.com/23504693/88766292-0ea20000-d178-11ea-81ac-df5e3cddcda1.png)

**Comment by hotcakedev628** [comment]
@arshaw How to customize `.fc-list-day` in `listview`?
![image](https://user-images.githubusercontent.com/60177750/103094235-dab5f400-45ca-11eb-864f-384facbad29e.png)

I was going to make `fc-list-day` to have two separated `<th>`, one is `<a class="fc-list-day-text">` and another one is `<a class="fc-list-day-side-text">`.
So I tried this, but it duplicates x2.

Here's my code.

```
         eventDidMount: function (arg) {
            $('.fc-list-day th').removeAttr("colspan");
            $('.fc-list-day').each(function (index, value) {
                var title_header = $(this).children('th').clone();
                $(title_header).insertAfter($(this).children('th').first());
               // And I was going to remove the elements which have the classes fc-list-day-side-text and fc-list-day-text from the two <th> elements.
            })
        },
```

I try this code, then it shows this as the result.(I have only two events)
![image](https://user-images.githubusercontent.com/60177750/103094442-7e9f9f80-45cb-11eb-8cbc-2bde12a0af19.png)

Is there only a way to hard code to make the custom event table header element?
And how to figure this out?

**Comment by costelxro** [comment]
Fullcalendar 5.4.1 May help someone 
This work for RESOURCES and EVENTS

```
resourceAreaColumns: [
{
  field: 'id',
  headerContent: 'ID'
},
{
  field: 'title',
  headerContent: 'Company Name'
}
],
// resources: 'your_json_feed_link',
resources: [
  {
	id: '2023',
	building: '460 Bryant A',
	title: 'Company name A',
	extendedProps: 
        {
	   isResourceHtml: 'YES',
	   resourceHtml: '<a href="#">Company name A</a>'
	}
   },
    {
	id: '2510',
	building: '54 Bryant B',
	title: 'Company name B',
	extendedProps: 
        {
	   isResourceHtml: 'YES',
	   resourceHtml: '<a href="#">Company name B</a>'
	}
   },
],
events: [
  {
	id: '2',
	resourceIds: ['2023'],
	title: 'Event title 1',
	start: '2021-02-09T10:30:00',
	end: '2021-02-09T11:30:00'
  },
  {
	id: '3',
	resourceIds: ['2510'],
	title: 'Event title 2',
	start: '2021-02-09',
	extendedProps: {
        	isHTML: 'YES',
		html: '<a href="#">Any HTML here</a>'
      	}
  },

// HERE IT IS...
	//FOR RESOURCES
resourceLabelContent: function(arg) {
	let spanEl  = document.createElement('span')
	let resHtml = arg.resource.extendedProps['resourceHtml'];

	if (arg.resource.extendedProps.isResourceHtml) {
		spanEl .innerHTML = resHtml
	} else {
		spanEl .innerHTML = arg.resource.title
	}
	let arrayOfDomNodesLabelContent = [ spanEl ]
	return { domNodes: arrayOfDomNodesLabelContent }
},
	//FOR EVENTS
eventContent: function(arg) {
	let divEl = document.createElement('div')
	let htmlTitle = arg.event._def.extendedProps['html'];
	if (arg.event.extendedProps.isHTML) {
		divEl.innerHTML = htmlTitle
	} else {
		divEl.innerHTML = arg.event.title
	}
	let arrayOfDomNodes = [ divEl ]
	return { domNodes: arrayOfDomNodes }
}
```

**Comment by dzero1** [comment]
Thanks 🙏  you save me @costelxro

**Comment by Edwardiv1** [comment]
This replaces the entire contents of the box. How would you replace only the 'title' that gets returned in the event data?

**Comment by akaspick** [comment]
@Edwardiv1 Agree, eventContent is an all or nothing approach.  I want to keep the time as-is and just make a title adjustment, but not possible with this approach unless I supply my own time content as well.

**Comment by modernben** [comment]
Avoiding a full replacement of the html, it looks like you can use the `eventDidMount` hook

**Warning jQuery ahead ;)**

```
eventDidMount: function(info) {
    $(info.el + ' .fc-event-title').html( $(info.el + ' .fc-event-title').text() );
},
```
